rfcs/text/0000_alert_actions_service.md

```diff
@@ -0,0 +1,210 @@
+
+In order to support multiple integrations with external services such as slack,
+email, pager-duty, twitter, etc. There is a need for an interface that an alert
+task would use to execute actions with customized content that is specific to
+that alert. In order to increase customization and extensibility there is a need
+for a way the task could fire an action without knowing the specifics of a
+particular service (i.e email, slack)
+
+# Detailed design
+
+This proposal defines one method to allow an alert task to be decoupled from the
+action(s) that the user wants the alert task to execute when the alert is
+*triggered*. The `action` is a preconfigured saved `instance` of a `actionType`
+for a specific `connectorType`
+
+## Executing Actions
+
+Each action instance can be fired with optional action type parameter. You need
+to have a defined action instance in order to fire an action.
+
+```JS
+server.actions.fire({
+  action: 'send message to slack',
+  actionType: 'notification',
+  params: {
+    message: 'This is a test message from kibana actions service',
+    title: 'custom title',
+    // destination: '<slack url>' // optional
+  },
+})
+```
+
+The action `send message to slack` is the instance that is being fired. See
+below for details how to create them
+
+## Creating Instances
+
+Instances are saved in the Kibana index and are used in order to `fire` an action with parameters. The
+`params` required are specific to the `actionType`. Changing the `actionType` also
+changes what `connectorTypes` are available for it.
+
+
+```JS
+server.actions.instance({
+  name: 'send message to slack',
+  actionType: 'notification',
+  connectorType: 'slack',
+  params: {
+    destination: '<slack url>',
+  },
+  connectorParams: {
+    channel: '#bot-playground',
+  },
+});
+```
+
+*Note: instances are saved objects that are space aware*
+
+## The Connectors
+
+The `connector` is what defines what happens when an action is *fired*. This
+will define the particular integration such as slack, email, etc.
+
+```JS
+server.actions.registerConnector({
+  actionType: 'notification',
+  connectorType: 'slack',
+  async handler({ params, connectorParams }) {
+    try {
+      const body = JSON.stringify({
+        ...connectorParams,
+        username: 'webhookbot',
+        text: params.message,
+        icon_emoji: ':ghost:',
+      });
+      const result = new url.URL(params.destination);
+      await fetch(result.toString(), {
+        method: 'POST',
+        body,
+      });
+    } catch (e) {
+      warn(`[slack] failed with ${e.message}`);
+    }
+  },
+});
+```
+
+## Defining Action Types
+
+Registering action types is the way that these actions are exposed to the
+alerts. Included in the definition are the required parameters that the alert
+should provide and those that are handed off to the *connectors* to fit into
+whatever service integration they provide.
+
+```JS
+server.actions.registerActionType({
+  name: 'notification',
+  fire: [
```

Comment:
I'd expect this to be named `params` similar to how it's used on the `fire` API call. Also could we leverage TypeScript here to define types for each of these fields? We should be able to use `keyof` and a few tricks to get type information for each param field.

Something like this might work:
```ts
type Param = string | number | boolean;
type Params = { [name: string]: Param };

interface ActionTypeParam<T extends Param> {
  optional?: boolean;
}

type ActionTypeParams<P extends Params> = {
  [K in keyof P]: ActionTypeParam<P[K]>
}
```

That may allow you to enforce the types when `actions.fire` is called by using overloads of the fire function (each plugin could extend this interface to add the required param types for any actionTypes they register.)

```ts
// Base type info in the actions service code itself
interface ActionFire {
  (name: string, params: ActionTypeParams<any>): Promise<MyReturnType>;
}

// Interface extended / merged by plugins that add new actionTypes
// For example, the alerting plugin could do this:
interface ActionFire {
  (name: 'send to slack', params: SendToSlackParams): Promise<MyReturnType>;
  (name: 'send email', params: SendEmailParams): Promise<MyReturnType>;
}
```

This may not be exactly what it looks like, but I believe this should be possible.

Reply:
I totally agree this should be named params, and was definitely thinking that we make the typescript interface enforce the param types in the fire function.